vms/platformvm/txs/mempool/mempool.go

```diff
@@ -4,42 +4,317 @@
+	avaxAssetID ids.ID
+
+	lock               sync.RWMutex
+	cond               *lock.Cond
+	tree               *btree.BTreeG[meteredTx]
+	txs                map[ids.ID]meteredTx
+	consumedUTXOs      *setmap.SetMap[ids.ID, ids.ID]
+	droppedTxIDs       *lru.Cache[ids.ID, error]
+	gasAvailable       gas.Gas
+	numTxsMetric       prometheus.Gauge
+	gasAvailableMetric prometheus.Gauge
 }
 
-func New(namespace string, registerer prometheus.Registerer) (*Mempool, error) {
-	metrics, err := txmempool.NewMetrics(namespace, registerer)
-	if err != nil {
-		return nil, err
+func New(
+	namespace string,
+	weights gas.Dimensions,
+	gasCapacity gas.Gas,
+	avaxAssetID ids.ID,
+	registerer prometheus.Registerer,
+) (*Mempool, error) {
+	numTxsMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "count",
+		Help:      "number of transactions in the mempool",
+	})
+
+	gasAvailableMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "gas_available",
+		Help:      "amount of gas available",
+	})
+
+	if err := errors.Join(
+		registerer.Register(numTxsMetric),
+		registerer.Register(gasAvailableMetric),
+	); err != nil {
+		return nil, fmt.Errorf("failed to register metrics: %w", err)
+	}
+
+	m := &Mempool{
+		weights:     weights,
+		avaxAssetID: avaxAssetID,
+		tree: btree.NewG[meteredTx](2, func(a, b meteredTx) bool {
+			if a.gasPrice != b.gasPrice {
+				return a.gasPrice < b.gasPrice
+			}
+
+			// Break ties with txID
+			return a.TxID.Compare(b.TxID) < 0
+		}),
+		txs:                make(map[ids.ID]meteredTx),
+		consumedUTXOs:      setmap.New[ids.ID, ids.ID](),
+		droppedTxIDs:       lru.NewCache[ids.ID, error](64),
+		gasAvailable:       gasCapacity,
+		numTxsMetric:       numTxsMetric,
+		gasAvailableMetric: gasAvailableMetric,
 	}
-	pool := txmempool.New[*txs.Tx](
-		metrics,
-	)
-	return &Mempool{Mempool: pool}, nil
+
+	m.cond = lock.NewCond(&m.lock)
+	return m, nil
 }
 
+// Add adds the tx to the mempool and clears its dropped status.
 func (m *Mempool) Add(tx *txs.Tx) error {
-	switch tx.Unsigned.(type) {
-	case *txs.AdvanceTimeTx:
-		return ErrCantIssueAdvanceTimeTx
-	case *txs.RewardValidatorTx:
-		return ErrCantIssueRewardValidatorTx
-	default:
-		return m.Mempool.Add(tx)
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.consumedUTXOs.HasOverlap(tx.InputIDs()) {
+		return mempool.ErrConflictsWithOtherTx
 	}
+
+	meteredTx, err := m.meter(tx)
+	if err != nil {
+		return fmt.Errorf("failed to meter tx: %w", err)
+	}
+
+	// Try to evict lower gas priced txs if we do not have enough remaining gas
+	// capacity
+	if meteredTx.gasUsed > m.gasAvailable {
+		if err := m.tryEvictTxs(meteredTx); err != nil {
+			return err
+		}
+	}
+
+	m.tree.ReplaceOrInsert(meteredTx)
+	m.txs[meteredTx.TxID] = meteredTx
+	m.consumedUTXOs.Put(meteredTx.TxID, meteredTx.InputIDs())
+	m.droppedTxIDs.Evict(meteredTx.TxID)
+	m.gasAvailable -= meteredTx.gasUsed
+
+	m.updateMetrics()
+	m.cond.Broadcast()
```

Comment:
we're signaling the sync.Cond, however we can never shrink the tree length in this method, can we?